src/python/pants/goal/completion.py

```diff
@@ -0,0 +1,184 @@
+
+        - By a user generating a completion script for their shell (e.g. `pants complete --zsh > pants-completions.zsh`)
+        - By the shell completions script when the user attempts a tab completion (e.g. `pants <tab>`, `pants fmt lint che<tab>`, etc...)
+
+        In the first case, we should generate a completion script for the specified shell and print it to stdout.
+        In the second case, we should generate the completion options for the current command and print them to stdout.
+
+        The trigger to determine which case we're in is the presence of the passthrough arguments. If there are passthrough
+        arguments, then we're generating completion options. If there are no passthrough arguments, then we're generating
+        a completion script.
+        """
+        if options._passthru:
+            completion_options = self._generate_completion_options(options)
+            if completion_options:
+                print("\n".join(completion_options))
+            return PANTS_SUCCEEDED_EXIT_CODE
+
+        script = self._generate_completion_script(self.shell)
+        print(script)
+        return PANTS_SUCCEEDED_EXIT_CODE
+
+    def _generate_completion_script(self, shell: Shell) -> str:
+        """Generate a completion script for the specified shell.
+
+        Implementation note: In practice, we're just going to read in
+        and return the contents of the appropriate static completion script file.
+
+        :param shell: The shell to generate a completion script for.
+        :return: The completion script for the specified shell.
+        """
+        if shell == Shell.ZSH:
+            return read_resource(_COMPLETIONS_PACKAGE, "pants-completion.zsh").decode("utf-8")
+        else:
+            return read_resource(_COMPLETIONS_PACKAGE, "pants-completion.bash").decode("utf-8")
+
+    def _generate_completion_options(self, options: Options) -> list[str]:
+        """Generate the completion options for the specified args.
+
+        We're guaranteed to have at least two arguments (`["pants", ""]`). If there are only two arguments,
+        then we're at the top-level Pants command, and we should show all goals.
+        - `pants <tab>` -> `... fmt fix lint list repl run test ...`
+
+        If we're at the top-level and the user has typed a hyphen, then we should show global options.
+        - `pants -<tab>` -> `... --pants-config-files --pants-distdir --pants-ignore ...`
+
+        As we add goals, we should show the remaining goals that are available.
+        - `pants fmt fix lint <tab>` -> `... list repl run test ...`
+
+        If there is a goal in the list of arguments and the user has typed a hyphen, then we should
+        show the available scoped options for the previous goal.
+        - `pants fmt -<tab>` -> `... --only ...`
+
+        :param options: The options object for the current Pants run.
+        :return: A list of completion options.
+        """
```

Comment:
most excellent docstring!! (best I've seen.) it made it so much easier to get into the implementation.

Reply:
I think we could consider returning early if we see `--` in our passthru args, as then we're in a position where pants doesn't care about what comes next.

Reply:
Note to self: Review this comment in https://github.com/pantsbuild/pants/issues/20676